### ISSUE TITLE
Warn users of inefficiency if they provide static indexing vectors as data

### ISSUE THREAD

**Issue by perrydv** [nimble-dev/nimble#951]
If a user is unclear on the distinction between data and constants, they might provide an indexing vector as data.  In that case, it will be interpreted as a node and therefore as a stochastic indexing situation, even if it is not.  This will create inefficiency due to larger dependency lists than necessary.  Do we want to identify such situations and warn users?

Reproducible example:
```{r}
library(nimble)

## Providing k as part of data and constants together when given as 
## constants.  This gives correct separation of data and constants, 
## with k treated as constant.
mc <- nimbleCode({
  for(i in 1:5) {
    x[i] ~ dnorm(0, 1)
  }
  for(i in 1:20) {
    y[i] ~ dnorm(x[ k[i] ], sd = 1)
  }
})

m <- nimbleModel(mc,
                 constants = list(x = rnorm(20),
                                  y = rnorm(20),
                                  k = rep(1:5, 4)))

m$getDependencies('x[1]') ## Correct

## Providing k in constants works correctly.
mc <- nimbleCode({
  for(i in 1:5) {
    x[i] ~ dnorm(0, 1)
  }
  for(i in 1:20) {
    y[i] ~ dnorm(x[ k[i] ], sd = 1)
  }
})

m <- nimbleModel(mc,
                 data = list(x = rnorm(20), ## x could go in constants
                             y = rnorm(20)),
                 constants = list(
                   k = rep(1:5, 4)))

m$getDependencies('x[1]') ## correct

## Providing k in data works but will give inefficient MCMC 
## (or whatever method) because it is handled as a stochastic indexing situation.
mc <- nimbleCode({
  for(i in 1:5) {
    x[i] ~ dnorm(0, 1)
  }
  for(i in 1:20) {
    y[i] ~ dnorm(x[ k[i] ], sd = 1)
  }
})

m <- nimbleModel(mc,
                 data = list(x = rnorm(20),
                             y = rnorm(20),
                             k = rep(1:5, 4)))

m$isData('k')
m$getDependencies('x[1]') ## all of y, as if there is stochastic indexing

```

**Comment by paciorek** [comment]
Yes, makes sense to warn. This situation would also arise if user provide 'k' as 'inits'.

I was thinking we could just check if `k[1],...,k[20]` appears in the nodes of the model, but this wouldn't catch a case where a user defines `k[i]` as a deterministic node. So to check that I think we'd need to check if `k[i]` has stochastic parents. Alternatively we could try to check if any RHSonly nodes are used in dynamic indexing. However I don't think we have stored info indicating node names for things used in dynamic indexing.

Maybe we should just check the basic case of whether `k[i]` is RHS only and assume that if a user is computing `k[i]` within model code that they have good reason for it.

However, if we go down that path, I'm still not seeing that we easily have all the pieces we need. 

`modelDefClass$findDynamicIndexParticipants()` generates this (for a more complicated case where we have `x[k[i]+1]`):

```
> m$modelDef$declInfo[[2]]$dynamicIndexInfo
[[1]]
[[1]]$indexCode
k[i] + 1

[[1]]$lower
[1] 1

[[1]]$upper
[1] 20
```

So we'd have to figure out that `k[i]` is the symbolic parent in the indexing and then figure out the actual nodes based on evaluating `i` and then check if they are RHS only. I don't think we have stored anywhere that k[1],...,k[20] are nodes used in a dynamic index expression. We just have stored that variable 'x' is dynamically indexed in `BUGSdecl$dynamicIndexInfo`.

**Comment by perrydv** [comment]
@paciorek, how about this:
```
m$modelDef$declInfo[[2]]$symbolicParentNodes
[[1]]
k[i]

[[2]]
.x_unknownIndex_x_oBk_oBi_cB_plus_1_cB_context3[NA_real_]
```

Is what we do to disentangle data from constants when both are provided as constants relevant?

**Comment by paciorek** [comment]
1) Not clear to me that is unambiguous. Consider that in this case it doesn't allow us to figure out if it's `j` or `k` that is involved in the indexing.

```
code <- nimbleCode({
    for(i in 1:3) {
        y[i] ~ dnorm(mu[k[i]+1]+j[i],1)
    }})

m = nimbleModel(code, inits = list(k = 1:3, j = rnorm(3), mu = rnorm(4)))
m$modelDef$declInfo[[1]]$symbolicParentNodes
> m$modelDef$declInfo[[1]]$symbolicParentNodes
[[1]]
k[i]

[[2]]
.mu_unknownIndex_mu_oBk_oBi_cB_plus_1_cB_context2[NA_real_]

[[3]]
j[i]
```

2) I don't think the data/constants disentangling is relevant. I think in this case we really need to be able to figure out what is used in indexing and then see if RHS only.  But at the time we are teasing out the dynamic indexes we are operating on the model code expressions and not the roles of nodes in the model...

**Comment by paciorek** [comment]
Ok, am working on a potential solution in fix_nct951_idx.

Note to self - check speed of the checking when have many nodes.

**Comment by paciorek** [comment]
Fixed by PR #963
